standalone-dns-proxy/cmd/standalonednsproxy.go

```diff
@@ -0,0 +1,578 @@
+				return err
+			}
+			log.WithField("policyState", policyState).Debug("Received policy state")
+
+			response := &standalonednsproxy.PolicyStateResponse{
+				RequestId: policyState.GetRequestId(),
+			}
+			revertStack, err := sdp.UpdatePolicyState(policyState)
+			if err != nil {
+				log.WithError(err).Error("Failed to update policy state")
+				revertStack.Revert()
+				err = sdp.policyStateStream.Send(response)
+				if err != nil {
+					log.WithError(err).Error("Failed to send policy state response")
+					return err
+				}
+				return err
+			}
+			response.Response = standalonednsproxy.ResponseCode_RESPONSE_CODE_NO_ERROR
+			err = sdp.policyStateStream.Send(response)
+			if err != nil {
+				log.WithError(err).Error("Failed to send policy state response")
+				return err
+			}
+		}
+	}
+}
+
+func (sdp *StandaloneDNSProxy) closePolicyStateStream() {
+	if sdp.policyStateStream != nil {
+		err := sdp.policyStateStream.CloseSend()
+		if err != nil {
+			log.Errorf("Failed to close policy state stream: %v", err)
+		}
+		sdp.policyStateStream = nil
+	}
+}
+
+func (sdp *StandaloneDNSProxy) closeConnection() error {
+	if sdp.connection != nil {
+		err := sdp.connection.Close()
+		if err != nil {
+			log.Errorf("Failed to close connection: %v", err)
+			return err
+		}
+		sdp.connection = nil
+	}
+	return nil
+}
+
+// createStreamPolicyStateStream creates a subscription stream to the policy state
+func (sdp *StandaloneDNSProxy) createStreamPolicyStateStream(ctx context.Context) error {
+	if sdp.Client == nil {
+		log.Error("Client is nil")
+		return fmt.Errorf("client is nil")
+	}
+
+	if sdp.policyStateStream != nil {
+		log.Error("Policy state stream is not nil")
+		sdp.closePolicyStateStream()
+	}
+
+	stream, err := sdp.Client.StreamPolicyState(ctx)
+	if err != nil {
+		log.WithError(err).Error("Failed to subscribe to policy state")
+		return err
+	}
+	sdp.policyStateStream = stream
+	return nil
+}
+
+// UpdatePolicyState updates the DNS rules in the standalone DNS proxy
+// 1. It updates the ip to identity cache and ip to endpoint id cache
+// 2. It updates the DNS rules in the standalone DNS proxy
+// The input is the policy state received from the cilium agent as :
+//
+//	PolicyState : {
+//	  EgressL7DnsPolicy : [
+//	    {
+//	    SourceEndpointId : 1
+//	    DnsServers : [{
+//	      DnsServerIdentity : 2
+//	      DnsServerPort : 53
+//	      DnsServerProto : 17
+//	    	},
+//	    	{
+//	      DnsServerIdentity : 3
+//	      DnsServerPort : 53
+//	      DnsServerProto : 17
+//	    	}]
+//	    DnsPattern : ["www.example.com"]
+//	    },
+//	    {
+//	    SourceEndpointId : 1
+//	    DnsServers : [{
+//	      DnsServerIdentity : 2
+//	      DnsServerPort : 54
+//	      DnsServerProto : 17
+//	    	}]
+//	    DnsPattern : ["www.test.com"]
+//	    }
+//	  ]
+//	  IdentityToEndpointMapping : []
+//	}
+func (sdp *StandaloneDNSProxy) UpdatePolicyState(rules *standalonednsproxy.PolicyState) (revert.RevertStack, error) {
```

Comment:
I'm not sure I understand why this is a snapshot based protocol. As far as I can tell, every change to a policy or an identity causes a full snapshot to be taken and copied over the proxy, which then also ingests the full snapshot. Why?

Reply:
The event of update/delete of policy is not being passed to the SDP. So as per the CFP we decided to send a snapshot of the policy state rather than adding complication of sending event type to SDP. 
In the case of snapshot, the SDP does not need to care about adding/deleting rules and can just depend(replace on its end) on the data cilium agent is sending without any manipulations.

The ingestion at the SDP side is using same proxy implementation as that of embedded dns proxy, SDP is currently reiterating over the snapshot to make it compliance with the internal definition of proxy.